src/interop-client.ts

```diff
@@ -0,0 +1,266 @@
+ * @returns `{ gwProvider, gwChainId }` ready to use.
+ * @throws If `env: 'local'` is selected without enough information to construct a provider.
+ * @example
+ * ```ts
+ * const { gwProvider, gwChainId } = resolveGateway(); // testnet by default
+ * const onMainnet = resolveGateway({ env: 'mainnet' });
+ * const local = resolveGateway({ env: 'local', gwRpcUrl: 'http://localhost:3250', gwChainId: 506n });
+ * ```
+ */
+export function resolveGateway(config: types.GatewayConfig = {}): {
+  gwProvider: JsonRpcProvider;
+  gwChainId: bigint;
+} {
+  if (
+    config.gwProvider &&
+    config.gwChainId !== null &&
+    config.gwChainId !== undefined
+  ) {
+    return {gwProvider: config.gwProvider, gwChainId: config.gwChainId};
+  }
+
+  const env = config.env ?? 'testnet';
+
+  // Presets
+  if (env === 'testnet' || env === 'mainnet') {
+    const preset = GATEWAY_PRESETS[env];
+    const gwChainId = config.gwChainId ?? preset.chainId;
+    const rpcUrl = config.gwRpcUrl ?? preset.rpcUrl;
+    const gwProvider = config.gwProvider ?? new JsonRpcProvider(rpcUrl);
+    return {gwProvider, gwChainId};
+  }
+
+  // Local
+  if (env === 'local') {
+    const gwProvider =
+      config.gwProvider ??
+      (config.gwRpcUrl ? new JsonRpcProvider(config.gwRpcUrl) : undefined);
+    if (
+      !gwProvider ||
+      config.gwChainId === null ||
+      config.gwChainId === undefined
+    ) {
+      throw new Error(
+        'Gateway config for env="local" requires gwRpcUrl and gwChainId (or explicit gwProvider + gwChainId).'
+      );
+    }
+    return {gwProvider, gwChainId: config.gwChainId};
+  }
+
+  throw new Error('Invalid gateway configuration');
+}
+
+export class InteropClient {
+  /** Resolved Gateway provider used for batch→GW block mapping. */
+  readonly gwProvider: JsonRpcProvider;
+  /** Resolved Gateway chain id used for root lookups on target chains. */
+  readonly gwChainId: bigint;
+
+  constructor(opts: {gateway?: types.GatewayConfig} = {}) {
+    const {gwProvider, gwChainId} = resolveGateway(opts.gateway);
+    this.gwProvider = gwProvider;
+    this.gwChainId = gwChainId;
+  }
+
+  /**
+   * Send a message via the L1Messenger on the source chain and return the `Sent` bundle.
+   *
+   * @param srcWallet - Wallet connected to the source L2.
+   * @param message   - Bytes or string; strings are UTF-8 encoded.
+   *
+   * @returns Sent — the minimal info needed to later prove inclusion.
+   */
+  async sendMessage(
+    srcWallet: Wallet,
+    message: ethers.BytesLike | string
+  ): Promise<types.Sent> {
+    const messenger = new ethers.Contract(
+      utils.L1_MESSENGER_ADDRESS,
+      utils.L1_MESSENGER,
+      srcWallet
+    );
+
+    const bytes =
+      typeof message === 'string' ? ethers.toUtf8Bytes(message) : message;
+
+    const tx = await messenger.sendToL1(bytes);
+    const rcpt = await (
+      await srcWallet.provider.getTransaction(tx.hash)
+    ).waitFinalize();
```

Comment:
Have you tested this on testnet/mainnet? how long does it take for the `waitFinalize()` to resolve? I don't know a different way to do this but wondering if this will take a lot of time to resolve...

Reply:
v29 is not on testnet yet just stage. We will wait to merge this until then so we know forsure the time it will take but I expect it to be longer. Locally, its not bad (~40 secs for me).